tensorflow_addons/image/threshold_ops.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Image thresholding ops."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+import numpy as np
+
+@tf.function
+def basic_threshold(image, threshold):
+	"""Applies a basic threshold operation and outputs a binary image.
+	
+	Args:
+	  image: A tensor of shape (num_rows, num_columns, num_channels) (HWC) or
+	  (num_rows, num_columns) (HW). The rank must be statically known (the
+	  shape is not `TensorShape(None)`.
+
+	  threshold: An integer value between 0 and 255 which is used to binarize the image by 
+	  comparing it with the pixel values.
+	
+	Returns:
+	  Binary thresholded image based on the threshold value 
+
+	Raises:
+	  TypeError: If `image` is an invalid type.
+	"""
+	with tf.name_scope(name or "basic_threshold"):
+		image = tf.convert_to_tensor(image, name="image")
+		
+		rank = image.shape.rank
+		if rank != 2 and rank != 3:
+			raise ValueError("Image should be either 2 or 3-dimensional.")
+ 
+		if not isinstance(threshold,int):
+			raise ValueError("Threshold value must be an integer.")
+
+		if rank == 3:
+			image = tf.image.rgb_to_grayscale(image)
+			image = tf.image.convert_image_dtype(image, tf.dtypes.uint8)
+			image = tf.squeeze(image,2)
+		
+		r, c = image.shape
+		image = image.numpy()
+		final = np.zeros((r,c))
+		final[image>threshold] = 255
+		return final
```

Comment:
I am afraid we have to make the codes compatible with graph mode. Is it possible to do these thing without `.numpy`?